### PR TITLE
Add support of Azure cross resource group nodes

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

Part of feature [Cross resource group nodes](https://github.com/kubernetes/features/issues/604).

This PR adds support of Azure cross resource group nodes that are labeled with `kubernetes.azure.com/resource-group=<rg-name>` and `alpha.service-controller.kubernetes.io/exclude-balancer=true`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

See designs [here](https://github.com/kubernetes/community/pull/2479).

**Release note**:

```release-note
Azure cloud provider now supports cross resource group nodes that are labeled with `kubernetes.azure.com/resource-group=<rg-name>` and `alpha.service-controller.kubernetes.io/exclude-balancer=true`
```

/sig azure
/kind feature

### CODE REVIEW

**Comment by feiskyer** [comment]
Ready for review now.

/assign @khenidak @andyzhangx @ddebroy

File: pkg/volume/util/util.go [hunk 1]
```diff
@@ -520,6 +520,10 @@ func GetPath(mounter volume.Mounter) (string, error) {
 // This means that a StatefulSet's volumes (`claimname-statefulsetname-id`) will spread across available zones,
 // assuming the id values are consecutive.
 func ChooseZoneForVolume(zones sets.String, pvcName string) string {
+	if zones.Len() == 0 {
```
Comment (ddebroy) [inline pkg/volume/util/util.go]: Thanks for finding and adding this! Can the same logic be also added to `ChooseZonesForVolume` below?

Comment (feiskyer) [reply pkg/volume/util/util.go]: Yep, of course

Comment (feiskyer) [reply pkg/volume/util/util.go]: Fixed in the new commits

File: pkg/cloudprovider/providers/azure/azure_vmss_cache.go [hunk 2]
```diff
@@ -27,15 +27,16 @@ import (
...
+	vmssCacheSeparator = "#"
 
 	nodeNameToScaleSetMappingKey = "k8sNodeNameToScaleSetMappingKey"
 	availabilitySetNodesKey      = "k8sAvailabilitySetNodesKey"
 
 	vmssCacheTTL                      = time.Minute
 	vmssVMCacheTTL                    = time.Minute
-	availabilitySetNodesCacheTTL      = 15 * time.Minute
-	nodeNameToScaleSetMappingCacheTTL = 15 * time.Minute
+	availabilitySetNodesCacheTTL      = 5 * time.Minute
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_vmss_cache.go]: any comment for this timing change?

File: pkg/cloudprovider/providers/azure/azure_vmss_cache.go [hunk 3]
```diff
@@ -49,19 +50,19 @@ func (ss *scaleSet) makeVmssVMName(scaleSetName, instanceID string) string {
 func extractVmssVMName(name string) (string, string, error) {
 	split := strings.SplitAfter(name, vmssNameSeparator)
 	if len(split) < 2 {
-		glog.Errorf("Failed to extract vmssVMName %q", name)
+		glog.V(3).Infof("Failed to extract vmssVMName %q", name)
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure_vmss_cache.go]: why change to info?

File: pkg/volume/azure_dd/azure_provision.go [hunk 4]
```diff
@@ -304,14 +304,16 @@ func (p *azureDiskProvisioner) Provision(selectedNode *v1.Node, allowedTopologie
 
 		if zoned {
 			// Set node affinity labels based on availability zone labels.
-			requirements := make([]v1.NodeSelectorRequirement, 0)
-			for k, v := range labels {
-				requirements = append(requirements, v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}})
-			}
+			if len(labels) > 0 {
```
Comment (andyzhangx) [inline pkg/volume/azure_dd/azure_provision.go]: is this code change related to this PR?

File: pkg/volume/util/util.go (hunk 1 above)
Comment (andyzhangx) [reply pkg/volume/util/util.go]: is this code change related to this PR?

File: pkg/cloudprovider/providers/azure/azure.go [hunk 5]
```diff
@@ -156,11 +159,16 @@ type Cloud struct {
 	metadata                *InstanceMetadata
 	vmSet                   VMSet
 
-	// Lock for access to nodeZones
-	nodeZonesLock sync.Mutex
+	// Lock for access to node caches
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure.go]: add more comments here, it locks: nodeZones, nodeResourceGroups, unmanagedNodes

File: pkg/cloudprovider/providers/azure/azure.go [hunk 6]
```diff
@@ -530,3 +569,76 @@ func (az *Cloud) GetActiveZones() (sets.String, error) {
...
 }
+
+// GetNodeResourceGroup gets resource group for given node.
+func (az *Cloud) GetNodeResourceGroup(nodeName string) (string, error) {
+	// Kubelet won't set az.nodeInformerSynced, always return configured resourceGroup.
+	if az.nodeInformerSynced == nil {
+		return az.ResourceGroup, nil
+	}
+
+	az.nodeCachesLock.Lock()
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure.go]: it's a get operation, is it necessary to put a lock here? I am worried about too many lock here

**Review by andyzhangx — CHANGES_REQUESTED** [review]
Some comments: 
- As there is no test to cover this feature now, could you also paste your manual test result
- It looks like some code change is not related to this PR?

**Comment by feiskyer** [comment]
> It looks like some code change is not related to this PR?

This is not true. Availability zones, managed disks, VMSS, VMAS are all supported for cross RG nodes.

> As there is no test to cover this feature now, could you also paste your manual test result

Following scenarios have been validated:

- Add new cross-RG VMSS nodes
- Add new cross-RG VMAS nodes
- Availability zones for such nodes
- Managed disks on such nodes

File: pkg/cloudprovider/providers/azure/azure.go [hunk 7]
```diff
@@ -530,3 +569,76 @@ func (az *Cloud) GetActiveZones() (sets.String, error) {
...
+}
+
+// GetUnmanagedNodes returns a list of nodes not managed by Azure cloud provider (e.g. on-prem nodes).
+func (az *Cloud) GetUnmanagedNodes() (sets.String, error) {
+	// Kubelet won't set az.nodeInformerSynced, always return nil.
+	if az.nodeInformerSynced == nil {
+		return nil, nil
+	}
+
+	az.nodeCachesLock.Lock()
```
Comment (andyzhangx) [inline pkg/cloudprovider/providers/azure/azure.go]: is it necessary to lock here since unmanagedNodes is a set, not map

**Comment by feiskyer** [comment]
@andyzhangx @ddebroy Moved fixes for ChooseZoneForVolume and ChooseZonesForVolume to another PR https://github.com/kubernetes/kubernetes/pull/67745, so that this PR is more clear and only focus on cross RGs features. PTAL

File: pkg/cloudprovider/providers/azure/azure.go (hunk 7 above)
Comment (feiskyer) [reply pkg/cloudprovider/providers/azure/azure.go]: It is still based on map.

**Review by andyzhangx — APPROVED** [review]
/lgtm
